### ISSUE TITLE
netty-tcnative incompatible with Conscrypt provider

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#6315]
When [using conscrypt](https://github.com/grpc/grpc-java/blob/master/SECURITY.md#bundling-conscrypt) like:
```java
Security.insertProviderAt(Conscrypt.newProvider(), 1);
```

On one machine we got lots of weird failures, but also:
```
java.lang.NoSuchFieldException
        at io.grpc.netty.shaded.io.netty.handler.ssl.OpenSslX509TrustManagerWrapper$3.run(OpenSslX509TrustManagerWrapper.java:123)
```

Source code for line: https://github.com/netty/netty/blob/netty-4.1.41.Final/handler/src/main/java/io/netty/handler/ssl/OpenSslX509TrustManagerWrapper.java#L123

It looks like Netty is trying to dig into the JDK to reuse code. But that is only used in ReferenceCountedOpenSslContext which is when Netty is using tcnative.

GrpcSslContexts [prefers tcnative](https://github.com/grpc/grpc-java/blob/ed845991dbb39401425d85e9aa6289afa5db5374/netty/src/main/java/io/grpc/netty/GrpcSslContexts.java#L228), which is how tcnative is being selected. Netty then tries to use Conscrypt because of [`getInstance("TLS")`](https://github.com/netty/netty/blob/netty-4.1.41.Final/handler/src/main/java/io/netty/handler/ssl/OpenSslX509TrustManagerWrapper.java#L154).

It's unclear what the best solution is here. We probably do want to figure out a fix for Netty to work with Conscrypt, but it's unclear what rules Netty should use, since it is already in hack territory. In the short-term, if Conscrypt is the provider for `"TLS"` we could select JDK+Conscrypt instead of tcnative in GrpcSslContext.

CC @normanmaurer

**Comment by voidzcy** [comment]
We probably should also hide the [two `configure(...)` methods](https://github.com/grpc/grpc-java/blob/56a410f5acee98d8b6b5d80e60196269adda2d66/netty/src/main/java/io/grpc/netty/GrpcSslContexts.java#L153-L211) (make them private). So, the choice of TLS provider becomes completely internal.

**Comment by ejona86** [comment]
@voidzcy, those two methods are necessary for people to select their provider and have the ALPN settings align properly. I don't see the purpose of removing them, and I think removing them would hamstring many use-cases, especially the ability to workaround issues in a provider.

**Comment by voidzcy** [comment]
This might not be a breakage for us, but might be a security issue.

The `NoSuchFieldException` error can never happen. The stacktrace is likely to be just a DEBUG log. Here is what Netty is doing: 

In JDK's implementation, the original X509TrustManager (created from the TrustManagerFactory with trust certs) is [wrapped with additional hostname verification](http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/cadea780bc76/src/share/classes/sun/security/ssl/SSLContextImpl.java#l1154) (becomes an X509ExtendedTrustManager) when the SSLContext is initialized. The hostname verification part is JDK's own logic. 
In order to match JDK's behavior (with hostname verification), Netty is trying to reuse JDK's code. It tries to find the wrapper class (a.k.a., [AbstractKeyManagerWrapper](http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/cadea780bc76/src/share/classes/sun/security/ssl/SSLContextImpl.java#l1312)) in JDK by [initiating a default SSLContext with dummy X509TrustManager](https://github.com/netty/netty/blob/7cf69022d48d123ac9f2920729613766f08f0874/handler/src/main/java/io/netty/handler/ssl/OpenSslX509TrustManagerWrapper.java#L61-L88) and see if wrapping happens. If it does, the original X509TrustManager can be passed into such an SSLContext to be wrapped with JDK's logic of hostname verification. However, the original X509TrustManager will not get wrapped (thus, no hostname verification) if the default SSLContext provider is not JDK (a.k.a., does not have the wrapping logic at SSLContext initialization). One example is when Conscrypt is inserted as the first security provider (Conscrypt performs hostname verification inside its own TrustManager implementation). This needs to be fixed in Netty, https://github.com/netty/netty/issues/10374 is created for it.

For us, when Conscrypt is present as the default SSLContext provider, JDK's wrapper code will not be found, OpenSslX509TrustManagerWrapper is just [no-op](https://github.com/netty/netty/blob/7cf69022d48d123ac9f2920729613766f08f0874/handler/src/main/java/io/netty/handler/ssl/OpenSslX509TrustManagerWrapper.java#L49). No hostname verification will be performed.

**Comment by ejona86** [comment]
netty/netty#10374 is fixed and will be part of Netty 4.1.51.Final. Since nobody has been calling out any problem here, I'm going to close this as fixed and we'll pull in the fix naturally as we upgrade Netty.

If a particular user is needing this, we can reopen it to track upgrading to 4.1.51+.
